File: app/app/src/main/java/me/williamhester/reddit/ui/fragments/VotableFragment.java
package me.williamhester.reddit.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;

import me.williamhester.reddit.messages.VotableListMessage;
import me.williamhester.reddit.models.Votable;
import me.williamhester.reddit.ui.adapters.VotableAdapter;

/** A fragment to show {@link Votable}s. */
public abstract class VotableFragment extends ContentFragment<VotableAdapter> {

  private final List<Votable> votables = new ArrayList<>();
  private boolean canLoad = true;

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    if (savedInstanceState == null) {
      loadContent();
    }
  }

  @Override
  public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    recyclerView.addOnScrollListener(new InfiniteLoadScrollListener());
  }

  @Override
  protected VotableAdapter createAdapter(Bundle savedInstanceState) {
    return new VotableAdapter(getLayoutInflater(savedInstanceState), htmlParser,
        contentClickCallbacks, this, getVotables());
  }

  @Override
  public void onVotableClicked(Votable votable) {
    contentClickCallbacks.navigateTo(votable);
  }

  @Subscribe(threadMode = ThreadMode.MAIN)
  public void receivedVotableList(VotableListMessage message) {
    setLoading(false);
    canLoad = true;
    int oldCount = getVotables().size();
    List<? extends Votable> newVotables = message.getVotables();
    if (newVotables.size() > 0) {
      getVotables().addAll(message.getVotables());
      adapter.notifyItemRangeInserted(oldCount, newVotables.size());
    }
  }

  protected final List<Votable> getVotables() {
    return votables;
  }

  @Nullable
  protected final String getAfter() {
    if (votables.size() == 0) {
      return null;
    }
    return votables.get(votables.size() - 1).getName();
  }

  private class InfiniteLoadScrollListener extends RecyclerView.OnScrollListener {
    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
      int position = layoutManager.findLastVisibleItemPosition();
      if (canLoad && position > getVotables().size() - 5) {
        canLoad = false;
        setLoading(true);
        loadContent();
      }
    }
  }
}

File: app/app/src/main/java/me/williamhester/reddit/ui/fragments/VotableFragment.kt
package me.williamhester.reddit.ui.fragments

import android.os.Bundle
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import me.williamhester.reddit.messages.VotableListMessage
import me.williamhester.reddit.models.Votable
import me.williamhester.reddit.ui.adapters.VotableAdapter
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.*

/** A fragment to show [Votable]s.  */
abstract class VotableFragment : ContentFragment<VotableAdapter>() {

  protected val votables = ArrayList<Votable>()
  private var canLoad = true

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    if (savedInstanceState == null) {
      loadContent()
    }
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    recyclerView.addOnScrollListener(InfiniteLoadScrollListener())
  }

  override fun createAdapter(savedInstanceState: Bundle?): VotableAdapter {
    return VotableAdapter(
        LayoutInflater.from(activity),
        htmlParser,
        contentClickCallbacks,
        this,
        votables)
  }

  override fun onVotableClicked(votable: Votable) {
    contentClickCallbacks.navigateTo(votable)
  }

  @Subscribe(threadMode = ThreadMode.MAIN)
  fun receivedVotableList(message: VotableListMessage) {
    isLoading = false
    canLoad = true
    val oldCount = votables.size
    val newVotables = message.votables
    if (newVotables.isNotEmpty()) {
      votables.addAll(message.votables)
      adapter.notifyItemRangeInserted(oldCount, newVotables.size)
    }
  }

  protected val after: String?
    get() {
      if (votables.size == 0) {
        return null
      }
      return votables[votables.size - 1].name
    }

  private inner class InfiniteLoadScrollListener : RecyclerView.OnScrollListener() {
    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
      // There's no way layoutManager can be null here.
      val position = layoutManager!!.findLastVisibleItemPosition()
      if (canLoad && position > votables.size - 5) {
        canLoad = false
        isLoading = true
        loadContent()
      }
    }
  }
}
